dcicutils/sheet_utils.py

```diff
@@ -0,0 +1,679 @@
+from typing import Any, Dict, Iterable, List, Optional, Tuple, Type, Union
+from .common import AnyJsonData, Regexp
+# from .env_utils import public_env_name, EnvUtils
+# from .ff_utils import get_schema
+from .lang_utils import conjoined_list, disjoined_list, maybe_pluralize  # , there_are
+from .misc_utils import ignored, pad_to, JsonLinesReader, remove_suffix  # , PRINT, AbstractVirtualApp
+# from .task_utils import pmap
+
+
+Header = str
+Headers = List[str]
+TabbedHeaders = Dict[str, Headers]
+ParsedHeader = List[Union[str, int]]
+ParsedHeaders = List[ParsedHeader]
+TabbedParsedHeaders = Dict[str, ParsedHeaders]
+SheetCellValue = Union[int, float, str]
+SheetRow = List[SheetCellValue]
+CsvReader = type(csv.reader(TemporaryFile()))
+SheetData = List[dict]
+TabbedSheetData = Dict[str, SheetData]
+JsonSchema = Dict
+TabbedJsonSchemas = Dict[str, JsonSchema]
+
+
+class LoadFailure(Exception):
+    """
+    In general, we'd prefer to load up the spreadsheet with clumsy data that can then be validated in detail,
+    but some errors are so confusing or so problematic that we need to just fail the load right away.
+    """
+    pass
+
+
+class LoadArgumentsError(LoadFailure):
+    """
+    Errors of this class represent situations where we can't get started because
+    there's a problem with the given arguments.
+    """
+    pass
+
+
+class LoadTableError(LoadFailure):
+    """
+    Errors of this class represent situations where we can't get started because
+    there's a problem with some table's syntax, for example headers that don't make sense.
+    """
+    pass
+
+
+def unwanted_kwargs(*, context, kwargs, context_plural=False, detailed=False):
+    if kwargs:
+        unwanted = [f"{argname}={value!r}" if detailed else argname
+                    for argname, value in kwargs.items()
+                    if value is not None]
+        if unwanted:
+            does_not = "don't" if context_plural else "doesn't"
+            raise LoadArgumentsError(f"{context} {does_not} use"
+                                     f" {maybe_pluralize(unwanted, 'keyword argument')} {conjoined_list(unwanted)}.")
+
+
+def prefer_number(value: SheetCellValue, kind='num'):
+    if isinstance(value, str):  # the given value might be an int or float, in which case just fall through
+        if not value:
+            return None
+        value = value
+        ch0 = value[0]
+        if ch0 == '+' or ch0 == '-' or ch0.isdigit():
+            if kind == 'num' or kind == 'int':
+                try:
+                    return int(value)
+                except Exception:
+                    pass
+            if kind == 'num' or kind == 'float':
+                try:
+                    return float(value)
+                except Exception:
+                    pass
+        # If we couldn't parse it as an int or float, fall through to returning the original value
+        pass
+    return value
+
+
+def expand_string_escape_sequences(text: str) -> str:
```

Comment:
Might be useful to docstring this since you're doing some fairly specific transformations

Reply:
Yeah, more doc strings are needed. This is a good one to make sure I do.